cowprotocol/bff
# feat(onchain-indexing): ✨ add indexing on cold start scenarios

## Warning
Please first review https://github.com/cowprotocol/cow-web-services/pull/7. This PR's size will be reduced once that is merged, as this branch is made off of `feature/orderbook` branch.

For actual diff, you can also look at https://github.com/cowprotocol/cow-web-services/compare/feature/orderbook-integration...feature/onchain-indexing

## What is this?
This PR adds on chain event indexing for TWAP orders. Changes here can be categorised as;
### utils
These are various utils to derive state out of ConditionalOrder data. In general, referring to cowswap project is useful, as they are derived from there.
- calculateValidTo: returns how long an order is valid, same as https://github.com/cowprotocol/cowswap/blob/develop/src/modules/twap/updaters/PartOrdersUpdater.tsx#L103
- computeOrderUid: returns order uid, used for generating order uid for order parts, same as https://github.com/cowprotocol/cowswap/blob/develop/src/utils/orderUtils/computeOrderUid.ts#L4
- generateOrderParts: returns/derives order parts for a given twap order, similar to https://github.com/cowprotocol/cowswap/commit/7901e1ac0fa6df6410be78382949675ac94be8a5
- getApiBaseUrl: return api base url for CoW api.
- getConditionalOrderId: derives conditional order id (order id of a twap), it has to be in sync with FE to be able to derive the same id for same order, same as https://github.com/cowprotocol/cowswap/blob/develop/src/modules/twap/utils/getConditionalOrderId.ts#L8
- getOrderStatus: derives order status, mainly;
  - if amount sold is same as total sell amount we expect (partSellAmount * n) => Fulfilled
  - if `singleOrder` is false => Cancelled. This is an array filled with booleans, as per https://github.com/rndlabs/composable-cow/blob/main/src/ComposableCoW.sol#L53, that is turned to false once an on chain cancellation happens as per https://github.com/rndlabs/composable-cow/blob/main/src/ComposableCoW.sol#L143C9-L143C9
  - if ConditionalOrderCreated event's block timestamp + timeInterval * amount of parts twap order has is in the past => Expired
  - if we don't have a block timestamp (not possible in this PR as we are only dealing with mined blocks) => WaitSigning
  - everything else => Pending
- parseOrderStruct: parses staticInput field of a twap order event (`event.args.params.staticInput`), has information defining the twap order such as amount of parts, sellAmount per part etc. https://github.com/cowprotocol/cowswap/blob/develop/src/modules/twap/utils/parseTwapOrderStruct.ts#L19
- transformers: added in #7 includes transformers to marshall / unmarshall database values.

### types
Currently only has `order.ts` here, all types in application should go here.

### migrations
Database migrations, as generated & adjusted afterwards via using the command `yarn run twap:generate-migrations`. As this PR includes database migrations, please don't forget to run migrations via `yarn run twap:run-migrations` (if you don't application will run it for you on first start, however still a good practice to do it yourself to see logs easier).

### libs/abis/src/generated/custom
This is an odd change, I am not happy with it, however can't find an alternative to exposing `ConditionalOrderCreatedEvent` otherwise. If you have a better option, please feel free to suggest it. It is useful to type events we get properly.

### plugins
`infura.ts` is where the fun happens here. This plugin essentially works in the following way;
- Per network, create an infura provider
- Fetch previously processed block
- Fetch all `ConditionalOrderCreated` events from this block until current block - some amount of lag (defined per chainId, at https://github.com/cowprotocol/cow-web-services/blob/9ce7b7243d032cefeafce38b46f77c3309778240/apps/twap/src/app/plugins/infura.ts#L20)
- Process these events with previous utilities mentioned, until we have a twap order structure we can rely on, and try to save this in a transaction with the current block number. This is an upsert operation, meaning if the order exists we just end up updating its value.

### data
We have updated `order.ts` here, to include a `status` field and to add a relation to a new entity `orderPart.ts`. Order part is as it sounds, the underlying orders of a twap, so that when you are processing a twap order you know which actual orders it related to.

We have also added `block.ts` which is a table to store blockNumbers we have processed.

## How to test?
Best way to test this is via connecting to your local database to see the data in there. To avoid inflating this PR, I'll add CRUD methods in a separate PR.
